### ISSUE TITLE
Expanding 2D fluxes to 3D

### ISSUE THREAD

**Issue by gmacgilchrist** [NOAA-GFDL/xwmt#3]
At present, 2D surface fluxes are  expanded to 3 dimensions so that their divergence can be computed in a manner consistent with 3D fluxes.

The present way of choosing to do this is to determine whether `lev_outer` is present in the dataset. See [here](https://github.com/jetesdal/xwmt/blob/a0a15a69801eaeed7f70b22ae95ba52325fbc479/xwmt/swmt.py#L108-L113). If it is, the variables are assumed to be 3D and will not be expanded, if it's not, they will be expanded.

However, it is possible for a Dataset to have `lev_outer` and still the fluxes as 2D - a Dataset only needs one variable to have the depth dimension for it to be present. For example, `rsdo` is a variable that is commonly posted with the surface fluxes. A "TODO" message related to this is present [here](https://github.com/jetesdal/xwmt/blob/a0a15a69801eaeed7f70b22ae95ba52325fbc479/xwmt/swmt.py#L114).

I need to give some thought to what is a more foolproof way to do this. The code loops through the variable anyway - can we instead just check at each variable whether it is 2 or 3 dimensional already, and expand the 2D ones?

**Comment by hdrake** [comment]
Another major issue with this is that the existing approach really only works for vertical coordinates for which the zeroth index is at the sea surface: 
```python
def expand_surface_to_3d(surfaceflux, z):
    """Expand 2D surface array to 3D array with zeros below surface"""
    return surfaceflux.expand_dims({"lev_outer": z}).where(z == z[0], 0)
```
Expanding 2D surface fluxes to outcropping `rho2` or `theta` levels, for example, sounds tricky...

**Comment by hdrake** [comment]
I agree that this is not a fundamental limitation, but I'm pretty sure the `.where(z == z[0], 0)` part of the current implementation will be wrong for, e.g., `rho2` coordinates where the surface fluxes are not in the zeroth index and thus will get masked to zero...

**Comment by hdrake** [comment]
> I need to give some thought to what is a more foolproof way to do this. The code loops through the variable anyway - can we instead just check at each variable whether it is 2 or 3 dimensional already, and expand the 2D ones?

This is what I do in https://github.com/NOAA-GFDL/xwmt/pull/40/commits/da06aea194d162d47538b8304c0c6d732a55a48e, see https://github.com/hdrake/xwmt/blob/da06aea194d162d47538b8304c0c6d732a55a48e/xwmt/wmt.py#L124-L138

Technically this would fail if there was a 2D file that had a single-valued vertical coordinate dimension, but I haven't come across that issue. It would be straightforward to modify the `bool` in the `if-else` block anyway.

**Comment by hdrake** [comment]
> At present, 2D surface fluxes are expanded to 3 dimensions so that their divergence can be computed in a manner consistent with 3D fluxes.

Following up on this, I don't think this is actually necessary. The whole reason for expanding the 2D fluxes to 3D is, as far as I can tell, for this step:
https://github.com/NOAA-GFDL/xwmt/blob/07b2e6fc95fd5bb715b461656509a8497c21b87a/xwmt/compute.py#L152-L160

However, I am pretty sure `Jlam` is exactly equal to `hlamdot` by definition, 
$h \dot{\lambda} \equiv \int_{z_{k}}^{z_{k}+h} - \nabla \cdot \mathcal{F}^{(z)} \text{d}z = -\mathcal{F_{\text{surf}}} \equiv \mathcal{J}_{\lambda}$ by the divergence theorem.

This means that we should be able to just get rid of this function altogether. However, this still does not address the fact that the surface flux needs to be added to the correct vertical coordinate level (trivial for `z` coordinates, but not obvious otherwise). I think this means we need to find the index of the outcropping layer.

**Comment by hdrake** [comment]
> I think this means we need to find the index of the outcropping layer.

A problem we'll have to deal with at a later date is how to find even find this for a non-monotonic coordinate. I guess just the vertical level whose value is closest to the local value of the 2D surface field (e.g. `tos`, `sos`)?

**Comment by hdrake** [comment]
> I think this means we need to find the index of the outcropping layer.

We also would need this for the ocean reference value is the surface mass flux formula $Q(\lambda_{m} - \lambda_{s})$

**Comment by hdrake** [comment]
PR https://github.com/NOAA-GFDL/xwmt/pull/40 addresses this issue by checking the dimensions of the tendency field:
https://github.com/hdrake/xwmt/blob/50e470b42b27a2c5bc6215f7b90c2c750a7bb320/xwmt/wmt.py#L187-L201

For 2D fields, we use the following methods in the `WaterMass` class to expand the 2D fluxes to 3D (outer levels) and to pick out the outcropping layer:
https://github.com/hdrake/xwmt/blob/50e470b42b27a2c5bc6215f7b90c2c750a7bb320/xwmt/wm.py#L154-L232
